### PR TITLE
get categories endpoint implemented

### LINKED ISSUE #38: (SP:  2) Implement "Get Categories" endpoint
## Description:

The objective of this task is to implement the backend functionality for retrieving categories  through a new endpoint. 
This will involve setting up the `route`, `controller`, `service` layer, and `tests`, as well as updating the `Swagger` documentation to reflect the new endpoint's capabilities. The implementation should adhere to best practices in terms of security, performance, and code readability

## Requirements:

- `Route:` 
  - Create a new route in the application to handle `GET` requests for fetching categories
- `Service` layer: 
  - Implement a service that will handle the business logic for retrieving  categories from the database
- `Controller:`
   - Develop a controller function to process incoming requests, validate query parameters, and return the appropriate response to the client. The `controller` should interact with the `service` layer to fetch the required data
- `Tests`: 
   - Write tests to ensure the endpoint functions correctly under various scenarios. Tests should cover successful data retrieval, handling of invalid requests
- `Swagger Documentation:` 
   - Update the `Swagger` docs to include the new `Get Categories` endpoint, providing clear information on the request parameters, response structure, and status codes. Ensure the documentation accurately reflects the endpoint's functionality and any query parameters it supports

`Middleware Integration:` Utilize existing middleware for authentication and role-based access control. 
Ensure that the endpoint is accessible only to authorized roles

`Error Handling:` Implement robust error handling to manage and respond to exceptions or validation errors, ensuring that the client receives meaningful error messages

`Code Quality:` Ensure the code is well-structured, commented, and follows project coding standards. Utilize helper functions and middleware as appropriate to keep the codebase `DRY` and maintainable

### PR DESCRIPTION
**Pull request by Annaivv**
- getCategories service added
- getCategories controller added
- route for getCategories added
- Swagger documentation for getCategories added
- unit tests for getCategories added

### CODE REVIEW

File: src/controllers/category.js [hunk 1]
```diff
@@ -1,11 +1,20 @@
 const categoryService = require('~/services/category')
 
+const getCategories = async (req, res) => {
+  const { skip, limit } = req.query
+
```
Comment (geersann) [inline src/controllers/category.js:7]: It might be useful to add a `NaN ` check for `skip` and `limit`

File: src/routes/category.js [hunk 2]
```diff
@@ -13,6 +13,31 @@ const params = [{ model: Category, idName: 'id' }]
...
+ *               $ref: '../swagger-schemas/category-schema.yaml#/components/schemas/Category'
+ *             example:
+ *               - _id: "66bd1977da1a3f609fe9a1af"
+ *                 name: "Language"
+ *                 appearance: {
+ *                      icon: "Language",
+ *                      color: "79b25f"
+ *                 }
+ *
+ */
```
Comment (geersann) [inline src/routes/category.js]: Probably, you can add `skip` and `limit` to swagger documentation

File: src/services/category.js [hunk 3]
```diff
@@ -1,6 +1,12 @@
 const Category = require('~/models/category')
 
 const categoryService = {
+  getCategories: async (skip = 0, limit = 10) => {
+    const items = await Category.find().skip(skip).limit(limit).exec()
+    const count = await Category.countDocuments()
```
Comment (geersann) [inline src/services/category.js]: Since we get the `item` and `count` at the same time, we can use the `PromiseAll` to reduce the waiting time

**Review by geersann — APPROVED** [review]

File: src/controllers/category.js (hunk 1 above)
Comment (Annaivv) [reply src/controllers/category.js:7]: Done [https://github.com/WebUI-Project-Based-10-nr/BackEnd/pull/88/commits/3bb7845a3ab4a57ceffdfdd8d7f13a635a52f761](url)

File: src/services/category.js (hunk 3 above)
Comment (Annaivv) [reply src/services/category.js]: Done [https://github.com/WebUI-Project-Based-10-nr/BackEnd/pull/88/commits/477ed579ab66ceb9ea7cc62c638a663e8849e8aa](url)

**Review by Yurii-04 — APPROVED** [review]

File: src/routes/category.js (hunk 2 above)
Comment (Annaivv) [reply src/routes/category.js]: Added [https://github.com/WebUI-Project-Based-10-nr/BackEnd/pull/88/commits/ccb4a67e47aa7de5f40a92ae76e39d50bd9dd8b4](url)

**Review by k7bart — APPROVED** [review]

File: src/controllers/category.js [hunk 4]
```diff
@@ -1,11 +1,21 @@
 const categoryService = require('~/services/category')
 
+const getCategories = async (req, res) => {
+  const skip = isNaN(parseInt(req.query.skip)) ? undefined : parseInt(req.query.skip)
+  const limit = isNaN(parseInt(req.query.limit)) ? undefined : parseInt(req.query.limit)
```
Comment (nrevutska) [inline src/controllers/category.js]: The ternary operator in this case is functional but could be simplified with a helper function like this
onst parseQueryInt = (value, defaultValue) => {
  const parsedValue = parseInt(value);
  return isNaN(parsedValue) ? defaultValue : parsedValue;
};
and then
const skip = parseQueryInt(req.query.skip, 0);
...

**Review by nrevutska — APPROVED** [review]

**Pull request merged**